### PR TITLE
Remove slavebook worker

### PR DESCRIPTION
**Pull request by elias19r**
For now, this is an experimental branch aimed to remove `SlaveBook` worker.
Related issue: https://github.com/rubykube/peatio/issues/2144

Some context:

- Every `Matching::Engine` has a `Matching::OrderBookManager` that owns two in-memory `Matching::OrderBook`: one for bid orders and one for ask orders
- Everytime an OrderBook is modified it calls its `broadcast` method that puts a payload to the `:slave_book` AMQP queue
- `SlaveBook` is a worker that listens to that queue and builds a copy of `Matching::Engine`s' `OrderBookManager` by reading the payloads it receives from the queue
- `SlaveBook` has a thread that writes its "depths" (a compact and accumulated version of its `OrderBookManager`s' `OrderBook PriceLevel`s) to `Rails.cache` on keys like `"peatio:#{market.id}:depths:{market.side}s"` every 3 seconds, even when nothing has been modified
- On the other hand, `global_state` daemon, every 5 seconds, triggers a publish event that reads depths from `Rails.cache` and sends and update message to users' websockets, also even when nothing has been modified.

Pull Request https://github.com/rubykube/peatio/pull/1894 rewrote `SlaveBook`'s `get_depth` method to fetch data from database making `SlaveBook`'s `OrderBookManager`s pointless and also the `:slave_book` queue and the `OrderBook.broadcast` method.

So the point is: can we remove `SlaveBook` worker and reduce the amount of data being sent to users?

This PR tries to do that by:
- Removing `SlaveBook` Worker
- Giving `Matching::OrderBook` the methods to build its "depths" and write them to `Rails.cache`
- Making `Matching::OrderBookManager`, responsible for checking if its `OrderBook`s have been modified and then trigger the publishing event

More specifically this PR:
- Refactors `Matching::OrderBook` to use mutexes to protect its `@limit_orders` and `@market_orders` (we'll see later there will be a concurrency)
- Replaces `Matching::OrderBook.broadcast` method by an `@updated` attribute that is set to true when some of its orders is modified (add/remove/update)
- Gives `build_depth` and `write_depth_to_cache` methods to `Matching::OrderBook`
- Gives `Matching::OrderBookManager` a thread that checks its `OrderBook`s for changes (every 500ms) and if so then it triggers the publish event (this thread reading from `OrderBook` competes with `Matching::Engine` modifying the `OrderBook`, so that's why the mutexes)
- A mandatory update can also be configured just in case (it was hardcoded every 30s)
- Add code to prevent leaking threads
- and finally, remove `SlaveBook` worker and its related stuff

So it is expected that depths will be written to `Rails.cache`, and users will receive update events on their websockets, only when some `OrderBook` has been updated.

I'd appreciate any suggestions and discussion on this topic :)

### CODE REVIEW

File: app/trading/matching/engine.rb [hunk 1]
```diff
@@ -17,6 +17,10 @@ def initialize(market, options={})
       shift_gears(options[:mode] || :run)
     end
 
+    def finalize
+      @orderbook.finalize
```
Comment (laka3000) [inline app/trading/matching/engine.rb:21]: You can use `delegate` here instead

**Comment by mod** [comment]
Hello, Thanks we noticed this issue, and what we did, is query the database from global_state.
You may not notice the change, but I will think @ysv will be able to show you.

**Comment by ysv** [comment]
@elias19r I will dig into your code.
Idea looks interesting for me.
We already did kind of hotfix for this [issue](https://github.com/rubykube/peatio/commit/125139cd6809c8c0d5665f4a71bb949c5bea24f1)

Thanks for your contribution to Peatio !
RubyKube team really appreciate your work

**Comment by huuhait** [comment]
nice work
i love it 👍

**Comment by elias19r** [comment]
Thanks @ysv

Yes, that hotfix from pull request https://github.com/rubykube/peatio/pull/1894 rewrote `SlaveBook.get_depth` method to fetch data from the database and removed SlaveBook's purpose which was to build a copy of `Matching::Engine`'s `OrderBookManager`s (through `:slave_book` queue) and serve a "depths" version of it to `Rails.cache`.

In this PR, I remove SlaveBook worker but managed to make `Matching::Engine`'s `OrderBookManagers` themselves to serve a "depths" to `Rails.cache`

**Comment by ysv** [comment]
@elias19r I had time to check your pr.
You did well but unfortunately it doesn't coincides with our view of solving this problem because:

- Your approach makes Matching::OrderBook slower

- Our plan is to support stop hidden orders which will appear in Matching::Orderbook but we will not show them for enduser so for the future reasons it's better to keep them in separate places

-  Also this approach makes complicated to read and understand Orderbook

For now we are kind of happy with single select per 3 second in SlaveBook but for sure we have this issue in mind and going to work on it soon.

Thanks for your contribution to Peatio !
RubyKube team really appreciate your work.
For now I'm closing this PR 
Put your comments in #2144

**Pull request closed without merging**